libraries/standard/mqtt/utest/mqtt_state_utest.c

```diff
@@ -0,0 +1,434 @@
+    TEST_ASSERT_EQUAL( status, MQTTBadParameter );
+    status = MQTT_ReserveState( NULL, 1, MQTTQoS1 );
+    TEST_ASSERT_EQUAL( status, MQTTBadParameter );
+
+    /* Test for collisions. */
+    mqttContext.outgoingPublishRecords[ 1 ].packetId = 1;
+    mqttContext.outgoingPublishRecords[ 1 ].qos = MQTTQoS1;
+    mqttContext.outgoingPublishRecords[ 1 ].publishState = MQTTPublishSend;
+
+    status = MQTT_ReserveState( &mqttContext, 1, MQTTQoS1 );
+    TEST_ASSERT_EQUAL( status, MQTTStateCollision );
+
+    /* Test for no memory. */
+    for( i = 0; i < MQTT_STATE_ARRAY_MAX_COUNT; i++ )
+    {
+        mqttContext.outgoingPublishRecords[ i ].packetId = 2;
+        mqttContext.outgoingPublishRecords[ i ].qos = MQTTQoS1;
+        mqttContext.outgoingPublishRecords[ i ].publishState = MQTTPublishSend;
+    }
+    status = MQTT_ReserveState( &mqttContext, 1, MQTTQoS1 );
+    TEST_ASSERT_EQUAL( status, MQTTNoMemory );
+
+    /* Success. */
+    resetPublishRecords( &mqttContext );
+    status = MQTT_ReserveState( &mqttContext, 1, MQTTQoS1 );
+    TEST_ASSERT_EQUAL( status, MQTTSuccess );
+    /* Reserve uses first available entry. */
+    TEST_ASSERT_EQUAL( mqttContext.outgoingPublishRecords[ 0 ].packetId, 1 );
+    TEST_ASSERT_EQUAL( mqttContext.outgoingPublishRecords[ 0 ].qos, MQTTQoS1 );
+    TEST_ASSERT_EQUAL( mqttContext.outgoingPublishRecords[ 0 ].publishState, MQTTPublishSend );
+}
+
+/* ========================================================================== */
+
+void test_MQTT_CalculateStatePublish( void )
+{
+    /* QoS 0. */
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_SEND, MQTTQoS0 ), MQTTPublishDone );
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_RECEIVE, MQTTQoS0 ), MQTTPublishDone );
+
+    /* QoS 1. */
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_SEND, MQTTQoS1 ), MQTTPubAckPending );
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_RECEIVE, MQTTQoS1 ), MQTTPubAckSend );
+
+    /* QoS 2. */
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_SEND, MQTTQoS2 ), MQTTPubRecPending );
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_RECEIVE, MQTTQoS2 ), MQTTPubRecSend );
+
+    /* Invalid QoS. */
+    TEST_ASSERT_EQUAL( MQTT_CalculateStatePublish( MQTT_SEND, 3 ), MQTTStateNull );
+}
+
+/* ========================================================================== */
+
+void test_MQTT_UpdateStatePublish( void )
+{
+    MQTTContext_t mqttContext = { 0 };
+    const uint16_t PACKET_ID = 1;
+    MQTTStateOperation_t operation = MQTT_SEND;
+    MQTTQoS_t qos = MQTTQoS0;
+    MQTTPublishState_t state;
+    int i = 0;
+
+    /* QoS 0. */
+    state = MQTT_UpdateStatePublish( &mqttContext, 0, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTPublishDone );
+
+    /* Invalid parameters. */
+    /* Invalid ID. */
+    qos = MQTTQoS1;
+    state = MQTT_UpdateStatePublish( &mqttContext, 0, operation, qos );
+    TEST_ASSERT_EQUAL( state, MQTTStateNull );
```

Comment:
This is more of a question regarding implementation. Why are we not using MQTTBadParameter in case of a bad parameter passed?

Reply:
We're tracking this issue separately (state engine refactor) for these functions to return status codes instead of the state.

Reply:
OK. Sounds good.